### PR TITLE
Re-organize using a proc-macro to support more devices

### PR DESCRIPTION
**Pull request by TethysSvensson**
# Summary

This is a continuation an continuation of the work I did in #716 and #719. It uses the same basic approach as #719, but I changed the macros in response to the feedback I got on that PR.

Since this is quite a large PR, I have split it up by commits, to make it easier to review.

A quick overview of what this PR contains:

* It adds a new proc-macro crate, which replaces all of the many-many features in the Cargo.toml
* The primary macro of that crate is `#[hal_cfg()]`, but it also contains a helper macro to allow using `#[hal_cfg()]` in more places, along with two specialized macros for creating conditional docstrings and conditional modules.
* A lot of doc-comments are currently conditional depending on feature flags. This PR removes most of these, as they cannot easily be supported by the new proc-macro.
* A lot of functionality were split up first by architecture in the `thumbv6m` and `thumbv7em` directories, and then by peripheral. This PR moves these files, so they are organized first by peripheral and then by architecture.
* The code for `reset_cause`, `serial_number` and `watchdog` were sufficiently similar that I decided to merge the implementations.
* It does not create a workspace, since I got weird crate unification errors when I did.

This PR also includes a change in `sercom/pad/impl_pad_thumbv6m.rs` should arguably be split into a separate PR.  The is that while converting the `pad_table`, I found a mistake in the table and did not feel like carrying that mistake over.

However I wanted to review the rest of the table for mistakes as well, which I found almost impossible to do unless I split up the tables by datasheet.

If you want me to split out this change in a separate PR, please let me know whether you prefer me to carry over the bug in the conversion, or if you prefer to fix the bug before merging this PR.

# Checklist
  - [x] `CHANGELOG.md` for the BSP or HAL updated
  - [x] All new or modified code is well documented, especially public items
  - [x] No new warnings or clippy suggestions have been introduced - CI will **deny** clippy warnings by default! You may `#[allow]` certain lints where reasonable, but ideally justify those with a short comment.

### CODE REVIEW

File: hal/src/sercom/i2c/flags.rs [hunk 1]
```diff
@@ -30,6 +30,12 @@ pub enum BusState {
     Busy = 0x03,
 }
 
+impl Default for Status {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
```
[the remark this answers is not in the document]
Comment (TethysSvensson) [reply hal/src/sercom/i2c/flags.rs:38]: Yep!

File: hal/src/dmac/channel/mod.rs [hunk 2]
```diff
@@ -356,6 +358,12 @@ pub enum CallbackStatus {
     TransferSuspended,
 }
 
+impl Default for InterruptFlags {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
```
[the remark this answers is not in the document]
Comment (TethysSvensson) [reply hal/src/dmac/channel/mod.rs:366]: To satisfy clippy

**Comment by jbeaurivage** [comment]
Nope it sounds good to me, I kinda like splitting the tables, it makes them more straightforward to understand. Anything else you want to review/discuss before we merge this?

**Comment by TethysSvensson** [comment]
I'm good with merging now. As I said I haven't done any automated backwards compatible testing.

I don't know when I'll have time to do so either, so if you want that done before merging I'd love some help.

**Comment by jbeaurivage** [comment]
Well, we rarely release patch updates anyway, so the next release will likely be a minor anyways, with no expectation of backwards compatibility. LGTM, thanks for the hard work!

**Pull request merged**